plugins/pulp_rpm/plugins/distributors/yum/publish.py

```diff
@@ -735,34 +735,41 @@ def _get_pkglist_to_publish(self, erratum_unit):
         seen_non_modular_packages = set()
         seen_modules = set()
         # find all pkglists for erratum despite the repo_id associated with pkglist
-        pkglists = models.ErratumPkglist.objects(errata_id=erratum_unit.errata_id)
-        for pkglist in pkglists:
-            for collection in pkglist['collections']:
-                module = collection.get('module')
-                if module:
-                    # inside a modular collection
-                    nsvca = self._module_dict_as_nsvca(module)
-                    if nsvca in seen_modules or nsvca not in self.repo_module_nsvca:
-                        # already processed or not from the repo being published
+        match_stage = {'$match': {'errata_id': erratum_unit.errata_id}}
+        unwind_collections_stage = {'$unwind': '$collections'}
+        unwind_packages_stage = {'$unwind': '$collections.packages'}
+        group_stage = {'$group': {'_id': '$collections.module',
+                                  'packages': {'$addToSet': '$collections.packages'}}}
+        collections = models.ErratumPkglist.objects.aggregate(
+            match_stage, unwind_collections_stage, unwind_packages_stage, group_stage,
+            allowDiskUse=True)
+
+        for collection in collections:
+            module = collection.get('_id')
```

Comment:
I would expect that this only returns the id, not the module data dict.  Is that not the case?  Everything below is the same as before, but it seems like the value of `module` has changed.

Reply:
In line 741. '_id' field is "$collections.module'  dict. For collection without module, '_id' field will be null.

Collection without module:
```json
> db.erratum_pkglists.aggregate([{'$match': {'errata_id': "RHSA-2021:1353"}}, {'$unwind': '$collections'}, {'$unwind': '$collections.packages'}, {'$group': {'_id': '$collections.module', 'packages': {'$addToSet': '$collections.packages'}}}], { allowDiskUse: true }).pretty()
{
	"_id" : null,
	"packages" : [
		{
			"src" : "thunderbird-78.10.0-1.el8_3.src.rpm",
			"name" : "thunderbird",
			"sum" : [
				"sha256",
				"c43ffb1772008b58de34fbdde6a8c0a61ece4b04a758a890e2ad25c202dfb633"
			],
			"filename" : "thunderbird-78.10.0-1.el8_3.x86_64.rpm",
			"epoch" : "0",
			"version" : "78.10.0",
			"release" : "1.el8_3",
			"arch" : "x86_64"
		}
	]
}
```

Collection with module:
```json
> db.erratum_pkglists.aggregate([{'$match': {'errata_id': "RHSA-2020:1644"}}, {'$unwind': '$collections'}, {'$unwind': '$collections.packages'}, {'$group': {'_id': '$collections.module', 'packages': {'$addToSet': '$collections.packages'}}}], { allowDiskUse: true }).pretty()
{
	"_id" : {
		"stream" : "10.6",
		"version" : "8020020200326162741",
		"arch" : "x86_64",
		"name" : "pki-core",
		"context" : "c7c3114f"
	},
	"packages" : [
		{
			"src" : "pki-core-10.8.3-1.module+el8.2.0+5925+bad5981a.src.rpm",
			"name" : "python3-pki",
			"sum" : [
				"sha256",
				"83ae27a7e8c83615639d080b34de3bf4f3a2c3808e819315f1cc22e063dba30f"
			],
			"filename" : "python3-pki-10.8.3-1.module+el8.2.0+5925+bad5981a.noarch.rpm",
			"epoch" : "0",
			"version" : "10.8.3",
			"release" : "1.module+el8.2.0+5925+bad5981a",
			"arch" : "noarch"
		},
		{
			"src" : "pki-core-10.8.3-1.module+el8.2.0+5925+bad5981a.src.rpm",
			"name" : "jss",
			"sum" : [
				"sha256",
				"779a8744572df84588864e2941095c83eada6dac6be42fa1fd244dc49beaa67b"
			],
			"filename" : "jss-4.6.2-4.module+el8.2.0+6123+b4678599.x86_64.rpm",
			"epoch" : "0",
			"version" : "4.6.2",
			"release" : "4.module+el8.2.0+6123+b4678599",
			"arch" : "x86_64"
		},
<snip>
```

Reply:
For any other pulp_rpm reviewers, note that all the changes below this are just an indentation change, the code is the same.

Reply:
yeah, I reviewed it as a part of BZ https://bugzilla.redhat.com/show_bug.cgi?id=1970246. Looks god to me.